Generalize TSIG keys and also CNAME and DNAME (#11)
DNS-Hackathon/catalog-extensions-draft: draft-axu-dnsop-catalog-zone-xfr-properties.md

### Signalling encrypted transports

The `primaries` property, *with* the extra label, MAY also have a TLSA RRset with one or more TLSA RRs {{!RFC6698}}.
The presence of a TLSA RRset signals support of DNS over TLS (DoT) {{!RFC7858}} or DNS over QUIC (DoQ) {{!RFC9250}} by the primary and the means by which the catalog consumer can successfully authenticate the primary.
TLSA RRsets MUST be prepended by two labels (below the `property` label *with* the extra label) indicating the decimal representation of the port number (see {{Section 3 of !RFC6698}}) and the protocol name of the transport (see {{Section 4 of ?I-D.draft-ietf-dnsop-svcb-dane-05}}).

> Catalog consumers MUST transfer member zone and incremental updates over either DoT or DoQ in the presence of a TLSA RRset.

Author: > I would prefer to let the consumer decide which security method to apply, also based on what the consumer actually supports.

I disagree. DoT and DoQ provide encryption (privacy) next to security, and TSIG only security. This would be similarly to how TLSA records can signal that a SMTP service MUST be done over TLS (See [Section 2.1.2. of RFC 7672](https://www.rfc-editor.org/rfc/rfc7672.html#section-2.1.2))

The key(s) MUST be defined elsewhere, such as in the configuration file of the consumer.
If the key cannot be found, the consumer MUST NOT notify the name server addresses for which the key was an additional attribute.
If the TXT RRset contains more than a single TXT RR, the consumer MUST NOT notify the name server addresses for which the key was an additional attribute.

~~~ ascii-art
notify.$CATZ                      0 IN A 192.0.2.49
notify.$CATZ                      0 IN TXT "name-of-default-key"

ZONELABEL3.zones.$CATZ            0 IN PTR example.org.
notify.ZONELABEL3.zones.$CATZ     0 IN AAAA 2001:db8:35::53
notify.ZONELABEL3.zones.$CATZ     0 IN TXT "keyname-for-ns4"

ns5.notify.ZONELABEL4.zones.$CATZ 0 IN AAAA 2001:db8:35::54
ns5.notify.ZONELABEL4.zones.$CATZ 0 IN TXT "keyname-for-ns5"
~~~

If there are any RRs attached to the `notify` property that are not covered by this document, they SHOULD be ignored.
 
## Allow Query

The `allow-query` property MAY be used to define an access list of hosts or networks that are allowed to send queries for the target zone(s).
The property MAY contain a RRset of type {{!RFC3123 (APL)}}, which MUST consist of only a single APL RR.

> The prefixes listed in the APL RR are processed in order:

Author: I believe not (See [Section 7 of RFC 3123](https://www.rfc-editor.org/rfc/rfc3123.html#section-7))

> The prefixes listed in the APL RR are processed in order:

Reviewer: Ah, I hadn't noticed that section. I think I'd like to hear from @0xebef about how Bind9 defines these semantics in their implementation.

> The prefixes listed in the APL RR are processed in order:

Reviewer 2: Excerpts from BIND 9 reference manual:

> The ACL is processed in order; if there is no
   match to any rule, the default policy is to deny access.

> When used as an access control list, a non-negated match allows access
and a negated match denies access. If there is no match, access is
denied.

> Order of insertion is significant. If more than one element in an ACL is
found to match a given IP address or prefix, preference is given to
the one that came *first* in the ACL definition. Because of this
first-match behavior, an element that defines a subset of another
element in the list should come before the broader element, regardless
of whether either is negated. For example, in ``1.2.3/24; ! 1.2.3.13;``
the 1.2.3.13 element is completely useless because the algorithm
matches any lookup for 1.2.3.13 to the 1.2.3/24 element. Using
``! 1.2.3.13; 1.2.3/24`` fixes that problem by blocking 1.2.3.13
via the negation, but all other 1.2.3.\* hosts pass through.

The key(s) MUST be defined elsewhere, such as in the configuration file of the consumer.

If an TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in transfers being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.
If an TXT RRset is present without an APL RRset, then only a TSIG key MUST match in any of the TXT RRs in the TXT RRset, regardless of the querying IP address.

If an `allow-transfer` property is present *and* contains APL RRsets and/or TXT RRsets (either directly below the property label or below the additional label), *and* none of the APLs and/or TSIG keys matched or could be found, then the consumer MUST NOT allow transfers of the member zone to which the property applies.

~~~ ascii-art
ZONELABEL5.zones.$CATZ  0                IN PTR example.local.
00-internal.allow-transfer.ZONELABEL5.zones.$CATZ 0 IN APL 1:10.0.0.0/8 2:fd00:0:0:0:0:0:0:0/8
50-external.allow-transfer.ZONELABEL5.zones.$CATZ 0 IN TXT "keyname"
~~~

# Implementation and Operational Notes

> The rationale for allowing CNAMEs for access lists is that a large and complex catalog may have large and complex access lists repeated a million times. But if there are only a few different access lists, they could be defined separately and then be referenced a million times, reducing both the size and processing effort of the catalog zone.

Reviewer: Sure, altho the point about size and processing effort is definitely an implementation note.
I think we should also add CNAMEs and DNAMEs in many of the example sections to drive the point in.

notify.ZONELABEL3.zones.$CATZ     0 IN TXT "keyname-for-ns4"

ns5.notify.ZONELABEL4.zones.$CATZ 0 IN AAAA 2001:db8:35::54
ns5.notify.ZONELABEL4.zones.$CATZ 0 IN TXT "keyname-for-ns5"
~~~

If there are any RRs attached to the `notify` property that are not covered by this document, they SHOULD be ignored.
 
## Allow Query

The `allow-query` property MAY be used to define an access list of hosts or networks that are allowed to send queries for the target zone(s).
The property MAY contain a RRset of type {{!RFC3123 (APL)}}, which MUST consist of only a single APL RR.
The prefixes listed in the APL RR are processed in order:
  - An IP address is allowed to query the zone when it matches a prefix.
  - An IP address is denied to query the zone when it matches a negated prefix.

The absence of an `allow-query` property at both apex of the catalog as well as at a member zone, means that the default policy applies, which may be that the member zone is allowed to be queried from any IP address without TSIG key.

Additional attributes (such as TSIG keys) can be bound to specific APL RRs by an additional label below the property label.

> The prefixes(with or without attributes) will be processed in {{Section 6 of !RFC4034 (canonical order)}}, which means that the RRsets at the `allow-query` property label will be processed first, followed by the RRsets with the additional label in canonical order.

Reviewer 2: Typo.

```suggestion
The prefixes (with or without attributes) will be processed in {{Section 6 of !RFC4034 (canonical order)}}, which means that the RRsets at the `allow-query` property label will be processed first, followed by the RRsets with the additional label in canonical order.
```

> An APL RRset containing more that a single APL RR, MUST be interpreted as an APL RRset containing a single APL RR denying all IP addresses, i.e.: `APL !1:0.0.0.0/0 !2:0:0:0:0:0:0:0:0/0`.

Author: ```suggestion
When a catalog consumer encounters an APL RRset containing more that a single APL RR, MUST be interpreted as an APL RRset containing a single APL RR denying all IP addresses, i.e.: `APL !1:0.0.0.0/0 !2:0:0:0:0:0:0:0:0/0`.
```

### TSIG Key Name

The `allow-query` property MAY also have a TXT RRset, which will (further) restrict the zone to be queryable only with the TSIG key indicated in any of the TXT RRs in the set.
The key(s) MUST be defined elsewhere, such as in the configuration file of the consumer.
A TXT RRset for a `allow-query` property containing more than a single TXT RR indicates a broken catalog zone that MUST NOT be processed (see {{Section 5.1 of !RFC9432}}).

> If an TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in queries being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.

Reviewer 2: Is the last part of this sentence suggesting that the TXT RRset can contain several TXT RRs? In the previous paragraph there is this:

> A TXT RRset for a `allow-query` property containing more than a single TXT RR indicates a broken catalog zone that MUST NOT be processed

Author: Yes, for the `primaries` and `notify` property it is important that there is only a single TSIG key defined, but for authenticating incoming connections (`allow-query` and `allow-transfer`), matching one in a set is also good.

I will remove the restriction of only a single TXT RR for TSIG Keys for the `allow-query` and `allow-transfer` properties.

> If an TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in queries being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.

Reviewer 2: Typo?

```suggestion
If a TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in queries being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.
```

> If an TXT RRset is present without an APL RRset, then only a TSIG key MUST match in any of the TXT RRs in the TXT RRset, regardless of the querying IP address.

Reviewer 2: Typo?

```suggestion
If a TXT RRset is present without an APL RRset, then only a TSIG key MUST match in any of the TXT RRs in the TXT RRset, regardless of the querying IP address.
```

~~~ ascii-art
ZONELABEL5.zones.$CATZ  0                IN PTR example.local.
00-internal.allow-query.ZONELABEL5.zones.$CATZ 0 IN APL 1:10.0.0.0/8 2:fd00:0:0:0:0:0:0:0/8
50-external.allow-query.ZONELABEL5.zones.$CATZ 0 IN TXT "keyname"
~~~

## Allow Transfer

The `allow-transfer` property MAY be used to define an access list of hosts or networks that are allowed to transfer the target zone(s) from the consumer.
The property MAY contain a RRset of type {{!RFC3123 (APL)}}, which MUST consist of only a single APL RR.
The prefixes listed in the APL are processed in order:
  - An IP address is allowed to query the zone when it matches a prefix.
  - An IP address is denied to query the zone when it matches a negated prefix.

The absence of an `allow-transfer` property at both apex of the catalog as well as at a member zone, signifies that transfers of the zone from the consumer MUST NOT be allowed.
Additional attributes (such as TSIG keys) can be bound to specific APL RRs by an additional label below the property label.

> The prefixes(with or without attributes) will be processed in {{Section 6 of !RFC4034 (canonical order)}}, which means that the RRsets at the `allow-transfer` property label will be processed first, followed by the RRsets with the additional label in canonical order.

Reviewer 2: Typo.

```suggestion
The prefixes (with or without attributes) will be processed in {{Section 6 of !RFC4034 (canonical order)}}, which means that the RRsets at the `allow-transfer` property label will be processed first, followed by the RRsets with the additional label in canonical order.
```

> An APL RRset containing more that a single APL RR, MUST be interpreted as an APL RRset containing a single APL RR denying all IP addresses, i.e.: `APL !1:0.0.0.0/0 !2:0:0:0:0:0:0:0:0/0`.

Reviewer 2: Same here - looks like defining the possibility that more than a single APL RR may exist in an APL RRset contradicts to a `MUST` restriction defined earlier:

> The property MAY contain a RRset of type {{!RFC3123 (APL)}}, which MUST consist of only a single APL RR.

### TSIG Key Name

The `allow-transfer` property MAY also have a TXT RRset, which will (further) restrict the zone to be transferable only with the TSIG key indicated in any of the TXT RRs in the set.
The key(s) MUST be defined elsewhere, such as in the configuration file of the consumer.
A TXT RRset for a `allow-transfer` property containing more than a single TXT RR indicates a broken catalog zone that MUST NOT be processed (see {{Section 5.1 of !RFC9432}}).

> If an TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in transfers being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.

Reviewer 2: Same question here about multiples TXT RRs in the TXT RRset.

Author: Resolved with commit 1e43b017

> If an TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in transfers being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.

Reviewer 2: Typo?

```suggestion
If a TXT RRset is present together with an APL RR, then first the policies in the APL are applied, and if that results in transfers being allowed for the IP address, then in addition a TSIG key MUST match any of the TXT RRs in the TXT RRset.
```

> If an TXT RRset is present without an APL RRset, then only a TSIG key MUST match in any of the TXT RRs in the TXT RRset, regardless of the querying IP address.

Reviewer 2: Typo?

```suggestion
If a TXT RRset is present without an APL RRset, then only a TSIG key MUST match in any of the TXT RRs in the TXT RRset, regardless of the querying IP address.
```

This section records the status of known implementations of the protocol defined by this specification at the time of posting of this Internet-Draft {{?RFC7942}}.

> The existing Bind9 implementation of `primaries`, `allow-transfer` and `allow-query` was a major inspiration for writing this draft.

Reviewer 2: ```suggestion
The existing BIND 9 implementation of `primaries`, `allow-transfer` and `allow-query` was a major inspiration for writing this draft.
```
